src/transactions/TransactionFrame.cpp

```diff
@@ -601,13 +641,36 @@ TransactionFrame::checkSorobanResources(
              makeU64SCVal(config.txMaxWriteBytes())});
         return false;
     }
-    if (readEntries.size() + writeEntries.size() >
-        config.txMaxDiskReadEntries())
+
+    uint32_t numDiskReads;
+    if (protocolVersionStartsFrom(ledgerVersion, AUTO_RESTORE_PROTOCOL_VERSION))
+    {
+        numDiskReads = getNumDiskReadEntries(resources, getResourcesExt());
+
+        auto totalReads = resources.footprint.readOnly.size() +
+                          resources.footprint.readWrite.size();
+        if (totalReads > config.txMaxInMemoryReadEntries())
+        {
+            diagnosticEvents.pushError(
+                SCE_STORAGE, SCEC_EXCEEDED_LIMIT,
+                "transaction total-entry-read resources exceed network config "
+                "limit",
+                {makeU64SCVal(totalReads),
+                 makeU64SCVal(config.txMaxInMemoryReadEntries())});
+            return false;
+        }
+    }
+    else
+    {
+        numDiskReads = readEntries.size() + writeEntries.size();
+    }
+
+    if (numDiskReads > config.txMaxDiskReadEntries())
     {
         diagnosticEvents.pushError(
             SCE_STORAGE, SCEC_EXCEEDED_LIMIT,
-            "transaction entry-read resources exceed network config limit",
-            {makeU64SCVal(readEntries.size() + writeEntries.size()),
+            "transaction entry-disk-read resources exceed network config limit",
```

Comment:
I think we should not change this error message for p22; it's ok to just have a protocol condition here and copy-paste the error text with the slight adjustment.